core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DeriveNotNull.java

```diff
@@ -0,0 +1,93 @@
+ * limitations under the License.
+ */
+package io.trino.sql.planner.iterative.rule;
+
+import com.google.common.collect.ImmutableSet;
+import io.trino.Session;
+import io.trino.matching.Captures;
+import io.trino.matching.Pattern;
+import io.trino.sql.ExpressionUtils;
+import io.trino.sql.planner.iterative.Rule;
+import io.trino.sql.planner.plan.JoinNode;
+import io.trino.sql.tree.Expression;
+import io.trino.sql.tree.IsNotNullPredicate;
+
+import java.util.Optional;
+
+import static io.trino.SystemSessionProperties.deriveIsNotNullPredicates;
+import static io.trino.sql.planner.plan.Patterns.join;
+import static io.trino.sql.tree.BooleanLiteral.TRUE_LITERAL;
+
+public class DeriveNotNull
+        implements Rule<JoinNode>
+{
+    // The purpose of this rule is to derive a conjunction of IS_NOT_NULL predicates from the join clauses for INNER joins.
+    // Although this may not directly benefit the current join node itself, the main goal is to provide the null-rejected condition for potential OUTER joins from upstream,
+    // and hopefully help with the transformation from outer joins to inner joins. Another benefit is possible cardinality reduction in upstream thanks to the derived predicates.
+
+    // Only apply this rule to inner join nodes, as IS_NOT_NULL predicates derived from inner join clauses won't change the semantic. That is not the case for outer joins.
+    private static final Pattern<JoinNode> PATTERN = join().matching(joinNode -> joinNode.getType() == JoinNode.Type.INNER && !joinNode.getCriteria().isEmpty());
+
+    @Override
+    public Pattern<JoinNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return deriveIsNotNullPredicates(session);
+    }
+
+    @Override
+    public Result apply(JoinNode node, Captures captures, Context context)
+    {
+        Expression filter = node.getFilter().isPresent() ? node.getFilter().get() : TRUE_LITERAL;
+        ImmutableSet<Expression> filterConjuncts = ImmutableSet.copyOf(ExpressionUtils.extractConjuncts(filter));
+
+        // In theory, we could derive "IS NOT NULL" information from both join conditions and filter conditions.
+        // For simplicity, we will only use join conditions for now.
+        // The rule can be extended later to support derivation from filter conditions.
+        for (JoinNode.EquiJoinClause joinClause : node.getCriteria()) {
+            Expression leftIsNotNullPredicate = new IsNotNullPredicate(joinClause.getLeft().toSymbolReference());
+            Expression rightIsNotNullPredicate = new IsNotNullPredicate(joinClause.getRight().toSymbolReference());
+
+            if (filterConjuncts.contains(leftIsNotNullPredicate)
+                    && filterConjuncts.contains(rightIsNotNullPredicate)) {
+                // this indicates we've matched and visited this node before
+                return Result.empty();
+            }
```

Comment:
If I understand correctly, this check along with inserting the `NOT NULL` conditions to the join filter, protects against falling into infinite loop.
This seems brittle. Since the `NOT NULL` predicates are pushed down, some other optimization (existing or future) could remove them from the join filter as redundant. Basically, this builds dependencies between rules, and we want to avoid that.  Ideally, the rules should not depend on each other at all.
Also, we don't want to evaluate the additional `NOT NULL` predicates in join.

I think that this change should be done in `PredicatePushDown`.

Reply:
Thanks @kasiafi for the review. Yes your understanding for that logic is correct - I want to avoid infinite loop by repeating this rule. But if I understand it correctly, other rules will only trigger after this rule finishes visiting all the applicable nodes. So there is no dependency on other rules and we should be safe here. Regarding the additional `NOT NULL` predicate, that's exactly what we want to generate, as it provides extra piece of information that's currently lacking in Trino to indicate the `null-rejected` property for one thing. Based on my benchmark tests I don't see any regression by this new rule.

Reply:
> other rules will only trigger after this rule finishes visiting all the applicable nodes

This is true as long as this rule is the only one in the `IterativeOptimizer`. We usually try to add new rules to existing optimizers. It is better explained here: https://github.com/trinodb/trino/pull/3688#discussion_r452859182
And my objection is that if the rule requires that the added `NOT NULL` conjuncts remain in the join filter (or it might loop), then we are forced to maintain it as a separate optimization step, or at least be cautious when adding other rules at the same step. Otherwise, some other rule/s might optimize-out the conjuncts from the join filter.

In my opinion, `PredicatePushDown` is the natural place for this change. It allows you to derive and push down the `NOT NULL` predicate without actually inserting it into the join filter.

BTW, there were attempts to add this optimization both in `PredicatePushDown` and as a rule. I think they both failed because of performance regressions due to evaluating the extra `NOT NULL` predicates.

Reply:
I see. Thanks for the explanation. The reason why I put it in a separate IterativeOptimizer was that I do see there are many such cases. But good to know we've moving toward a better direction of merging those as much as possible.